File: cpp/2021_합승택시요금.cpp
#include <string>
#include <vector>      
#include <queue>
#include <algorithm>
using namespace std;
#define INF 1e9 

vector<pair<int, int> > graph[100001];
int d1[100001]; 
int d2[100001]; 
void dijkstra(int start,int d[])
{
    priority_queue<pair<int,int>>pq;
    pq.push({0,start}); 
    d[start]=0;
    while(!pq.empty()) 
    {
        int dist = -pq.top().first; 
        int now = pq.top().second; 
        pq.pop();
        if(d[now]<dist)
            continue;
        for(int i=0; i<graph[now].size(); i++)
        {
            int cost = dist+graph[now][i].second;                                      
            if(cost<d[graph[now][i].first]) {
                d[graph[now][i].first]=cost;
                pq.push(make_pair(-cost,graph[now][i].first));
            }
        }
    }
}
int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int answer = 0;
    int noRide=0;
    vector <int> d4;
    fill(d1, d1 + 100001, INF);
    fill(d2, d2 + 100001, INF);
    for(long i=0;i<fares.size();i++){
        graph[fares[i][0]].push_back({fares[i][1], fares[i][2]});
        graph[fares[i][1]].push_back({fares[i][0], fares[i][2]});
    }
    dijkstra(s,d1);
    noRide+=d1[a];
    noRide+=d1[b];
    for(int i=1;i<=n;i++){
        fill(d2, d2 + 100001, INF);
        dijkstra(i,d2);
        d4.push_back(d1[i]+d2[a]+d2[b]);
    }
    int minNum=INF;
    for(int i=0;i<d4.size();i++){
        if(minNum>d4[i]&&d4[i]>0)
            minNum=d4[i];
    }
    answer=min(minNum,noRide);
    return answer;
}

File: python/2021_합승택시요금.py
import heapq 
from collections import defaultdict

def solution(n, s, a, b, fares):
    dis=defaultdict(list)
    result=[0 for i in range(n)]
    #딕셔너리 초기화
    for a1,b1,c1 in fares:
        dis[str(a1)].append([str(b1),c1])
        dis[str(b1)].append([str(a1),c1])
    #다익스트라 적용
    for i in range(1,n+1):
        result[i-1]=(dijkstra(dis,str(i)))

    #최소값 
    minNum=float('inf')
    for key in result[s-1]:
        num=result[s-1][key]+result[int(key)-1][str(a)]+result[int(key)-1][str(b)]
        if minNum>=num:
            minNum=num
    return minNum

def dijkstra(graph, start):
    distances = {node: float('inf') for node in graph}  #무한대로 초기화 
    distances[start] = 0  # 시작 값은 0이어야 함
    queue = []
    heapq.heappush(queue, [distances[start], start])  

    while queue:  
        current_distance, current_destination = heapq.heappop(queue)  

        if distances[current_destination] < current_distance:  
            continue
    
        for new_destination, new_distance in graph[current_destination]:
            distance = current_distance + new_distance  
            if distance < distances[new_destination]: 
                distances[new_destination] = distance
                heapq.heappush(queue, [distance, new_destination])  
    
    return distances